File: python/sample/virtual_call.py
import etsm

class FooState(etsm.State):
    def __init__(self, enter, exit, tick):
        super().__init__(enter, exit)
        self.Tick = tick

class Foo:
    def __init__(self):
        self.__sm = etsm.StateMachine(self)
        self.__a = FooState(None, None, Foo.TickA)
        self.__b = FooState(None, None, Foo.TickB)

    def TickA(self):
        print(' A ', end='')

    def TickB(self):
        print(' B ', end='')

    # call tick on the current state, aka virtual call
    def Tick(self):
        if ( self.__sm.Current != None ):
            self.__sm.Current.Tick(self)

    def Run(self):
        self.Tick();
        self.__sm.Transition(self.__a);
        self.Tick();
        self.__sm.Transition(self.__b);
        self.Tick();
        self.__sm.Transition(None);
        self.Tick();

foo = Foo()
foo.Run()

File: ruby/sample/virtual_call.rb
require 'etsm.rb'

class FooState < Etsm::State
    attr_reader :tick

    def initialize(enter, exit, tick)
        super(enter, exit)
        @tick = tick
    end
end

class Foo
    def initialize
        @sm = Etsm::StateMachine.new(self)
        @a = FooState.new(nil, nil, self.method(:TickA))
        @b = FooState.new(nil, nil, self.method(:TickB))
    end

    def TickA
        print ' A '
    end

    def TickB
        print ' B '
    end

    def Tick
        if @sm.current != nil and @sm.current.tick != nil
            @sm.current.tick.call
        end
    end

    def Test()
        self.Tick
        @sm.Transition(@a)
        self.Tick
        @sm.Transition(@b)
        self.Tick
        @sm.Transition(nil)
        self.Tick
    end
end

foo = Foo.new
foo.Test() # Output: " A  B "

